### ISSUE TITLE
[BUG] v0.9.4 zero3 `deepspeed.zero.GatheredParameters` failed to work

### ISSUE THREAD

**Issue by nrailgun** [microsoft/DeepSpeed#3836]
**Describe the bug**

Alpaca start with too large loss in v0.9.4.

Alpaca initializes word embedding like this:

```python
def smart_tokenizer_and_embedding_resize(
    special_tokens_dict: Dict,
    tokenizer: transformers.PreTrainedTokenizer,
    model: transformers.PreTrainedModel,
):
    """Resize tokenizer and embedding.

    Note: This is the unoptimized version that may make your embedding size not be divisible by 64.
    """
    num_new_tokens = tokenizer.add_special_tokens(special_tokens_dict)
    model.resize_token_embeddings(len(tokenizer))

    if num_new_tokens > 0:
        input_embeddings = model.get_input_embeddings().weight.data
        output_embeddings = model.get_output_embeddings().weight.data

        input_embeddings_avg = input_embeddings[:-num_new_tokens].mean(dim=0, keepdim=True)
        output_embeddings_avg = output_embeddings[:-num_new_tokens].mean(dim=0, keepdim=True)

        input_embeddings[-num_new_tokens:] = input_embeddings_avg
        output_embeddings[-num_new_tokens:] = output_embeddings_avg
```

I thought it's alpaca's bug, since it forgot to use `deepspeed.zero.GatherParameters` as mentioned in the [doc](https://www.deepspeed.ai/tutorials/zero/#allocating-massive-megatron-lm-models), but I found it doesn't help.

**To Reproduce**
Just train [standford_alpaca](https://github.com/tatsu-lab/stanford_alpaca).
```
torchrun --nproc_per_node=8 --master_port=12345 train.py \
    --model_name_or_path /data/nrwu/model-ckpt/llama-7b-ckpt-hf/ \
    --data_path ./alpaca_data.json \
    --bf16 True \
    --output_dir output-7B \
    --num_train_epochs 3 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 8 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 2000 \
    --save_total_limit 1 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --deepspeed "./configs/default_offload_opt_param.json" \
    --tf32 True \
    --logging_steps 1 \
    --report_to none
```

**Expected behavior**
The loss should start around 2.x.

**ds_report output**
```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
async_io ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.0
 [WARNING]  using untested triton version (2.0.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/pretrainx/conda/lib/python3.10/site-packages/torch']
torch version .................... 2.0.1+cu117
deepspeed install path ........... ['/home/pretrainx/conda/lib/python3.10/site-packages/deepspeed']
deepspeed info ................... 0.9.1, unknown, unknown
torch cuda version ............... 11.7
torch hip version ................ None
nvcc version ..................... 11.7
deepspeed wheel compiled w. ...... torch 2.0, cuda 11.7
```

**Screenshots**
The loss should start around 2.x (tested with DS v0.9.1).
![image](https://github.com/microsoft/DeepSpeed/assets/14273895/10a899d7-9432-4bd4-acd7-6b7e1d7b7e9e)

But actually it start around 10.x (tested with DS v0.9.4).
![image](https://github.com/microsoft/DeepSpeed/assets/14273895/12aff86b-1e8b-4da2-84cb-c0c5a3e403cc)

**System info (please complete the following information):**
 - OS: CentOS 7.x
 - 8 x A100 40GB
 - NvLink
 - Py 3.10

**Launcher context**
Are you launching your experiment with the `deepspeed` launcher, MPI, or something else?
```
torchrun --nproc_per_node=8 --master_port=12345 train.py \
    --model_name_or_path /data/nrwu/model-ckpt/llama-7b-ckpt-hf/ \
    --data_path ./alpaca_data.json \
    --bf16 True \
    --output_dir output-7B \
    --num_train_epochs 3 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 8 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 2000 \
    --save_total_limit 1 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --deepspeed "./configs/default_offload_opt_param.json" \
    --tf32 True \
    --logging_steps 1 \
    --report_to none
```

**Docker context**
```dockerfile
FROM geminihub.woa.com/yard/tlinux2-py3.6-cuda11.0
LABEL maintainer="pretrainx"

ARG MY_HOME=/home/pretrainx
WORKDIR ${MY_HOME}
RUN adduser -ms /bin/bash -d /home/pretrainx -g root pretrainx

# upgrade to cuda 11.7
RUN wget 'https://developer.download.nvidia.com/compute/cuda/11.7.0/local_installers/cuda_11.7.0_515.43.04_linux.run' -O cuda.run \
  && chmod +x cuda.run \
  && ./cuda.run --silent --toolkit \
  && rm cuda.run

# upgrade gcc4 to gcc9
RUN rpm --rebuilddb && yum -y install tlinux-release-scl && yum -y install scl-utils && yum install -y devtoolset-9-gcc devtoolset-9-gcc-c++
SHELL [ "/usr/bin/scl", "enable", "devtoolset-9"]

# install rdma driver
RUN rpm --rebuilddb && yum install -y redhat-rpm-config rpm-build libnl3 tk
RUN wget 'https://content.mellanox.com/ofed/MLNX_OFED-5.8-2.0.3.0/MLNX_OFED_LINUX-5.8-2.0.3.0-rhel7.2-x86_64.tgz' -O mlnxofed.tgz \
  && tar -xvf mlnxofed.tgz \
  && cd 'MLNX_OFED_LINUX-5.8-2.0.3.0-rhel7.2-x86_64' \
  && ./mlnxofedinstall --skip-distro-check --distro 'rhel7.2' --user-space-only --force \
  && cd .. \
  && rm -rf mlnxofed.tgz 'MLNX_OFED_LINUX-5.8-2.0.3.0-rhel7.2-x86_64'


# openmpi
RUN mkdir /tmp/openmpi && cd /tmp/openmpi && \
  wget 'https://download.open-mpi.org/release/open-mpi/v4.1/openmpi-4.1.2.tar.gz' && \
  tar zxf 'openmpi-4.1.2.tar.gz' && \
  cd 'openmpi-4.1.2' && \
  ./configure --enable-orterun-prefix-by-default --disable-mpi-fortran --enable-static && \
  make -j 8 all && \
  make install && \
  ldconfig && \
  rm -rf /tmp/openmpi

# miniconda
RUN wget 'https://repo.anaconda.com/miniconda/Miniconda3-py310_23.3.1-0-Linux-x86_64.sh' -O /home/pretrainx/miniconda3.sh \
  && bash ./miniconda3.sh -b -p ${MY_HOME}/conda \
  && echo 'export PATH=/home/pretrainx/conda/bin:${PATH}' >>/root/.bashrc \
  && rm ./miniconda3.sh
ENV PATH=${MY_HOME}/conda/bin:${PATH}
RUN python3 -m pip install --no-cache-dir --upgrade pip

RUN echo 'source scl_source enable devtoolset-9' >> /etc/bashrc

# pytorch
RUN python3 -m pip install 'torch==2.0.1+cu117' torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu117

CMD ["/bin/bash"]
```

```dockerfile
# syntax=docker/dockerfile:1
FROM docker.yard.oa.com:14917/pretrainx/base:v0.6.2

ARG MY_HOME=/home/pretrainx
WORKDIR ${MY_HOME}

RUN python3 -m pip install packaging
RUN rpm --rebuilddb && yum -y install ninja-build
RUN MAX_JOBS=8 mkdir -p /tmp/apex \
  && cd /tmp/apex \
  && git clone --recursive https://github.com/NVIDIA/apex \
  && cd apex \
  && git checkout 8b7a1ff183741dd8f9b87e7bafd04cfde99cea28 \
  && git submodule update --init --recursive \
  && python3 -m pip install --global-option="--cpp_ext" --global-option="--cuda_ext" --no-cache -v --disable-pip-version-check . \
  && cd /tmp \
  && rm -rf /tmp/apex

RUN MAX_JOBS=8 mkdir -p /tmp/flash-attn \
  && cd /tmp/flash-attn \
  && git clone --recursive https://github.com/HazyResearch/flash-attention.git \
  && cd flash-attention \
  && git checkout 7c766b1bbc65d341f5f7fcead82946319c9657d0 \
  && git submodule update --init --recursive \
  && python3 setup.py install \
  && cd /tmp/flash-attn/flash-attention/csrc/rotary \
  && python3 setup.py install \
  && cd /tmp/flash-attn/flash-attention/csrc/layer_norm \
  && python3 setup.py install \
  && cd /tmp \
  && rm -rf /tmp/flash-attn

RUN CC='/usr/local/bin/mpicc' python3 -m pip install --no-cache-dir "mpi4py==3.1.4"
RUN python3 -m pip install "pybind11==2.10"
RUN python3 -m pip install "six==1.16"
RUN python3 -m pip install "regex==2023.6.3"
RUN python3 -m pip install "numpy==1.24"
RUN python3 -m pip install "python-config==0.1"
RUN python3 -m pip install "datasets==2.12"
RUN python3 -m pip install "einops==0.6"
RUN python3 -m pip install "nltk==3.8"
RUN python3 -m pip install "accelerate==0.20.3"
RUN python3 -m pip install "tabulate==0.9"
RUN python3 -m pip install "transformers==4.30.2"
RUN python3 -m pip install "statsmodels==0.14"
RUN python3 -m pip install "tiktoken==0.3.2"
RUN python3 -m pip install "markdown==3.4"
RUN python3 -m pip install "scikit-learn==1.2"
RUN python3 -m pip install "sentencepiece==0.1.99"
RUN python3 -m pip install "pandas==2.0"
RUN python3 -m pip install "wandb==0.15"
RUN python3 -m pip install "fire==0.5"
RUN python3 -m pip install "openai==0.27"
RUN python3 -m pip install "protobuf==3.20"
RUN python3 -m pip install "tqdm==4.65"
RUN python3 -m pip install "evaluate==0.4"
RUN python3 -m pip install "loralib==0.1"
RUN python3 -m pip install "tokenizers==0.13"
RUN python3 -m pip install "fastapi==0.97"
RUN python3 -m pip install "sse_starlette==1.6"
RUN python3 -m pip install "deepspeed==0.9.4"
RUN python3 -m pip install "peft==0.3.0"
RUN python3 -m pip install "scipy==1.10"
RUN python3 -m pip install "langchain"
RUN python3 -m pip install "py-spy"
```

**Comment by jomayeri** [comment]
Please rerun with the latest version of DeepSpeed. If the issue occurs again please open another issue. Thanks.
